src/Passes/Source/Rules/OperandPrototype.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Rules/OperandPrototype.hpp"
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    IOperandPrototype::~IOperandPrototype() = default;
+    bool IOperandPrototype::matchChildren(Value* value, Captures& captures) const
+    {
+        auto user = llvm::dyn_cast<llvm::User>(value);
+        if (!children_.empty())
+        {
+            if (user == nullptr)
+            {
+                return false;
+            }
+
+            if (user->getNumOperands() != children_.size())
+            {
+                return false;
+            }
+
+            uint64_t i = 0;
+            while (i < children_.size())
+            {
+                auto v = user->getOperand(static_cast<uint32_t>(i));
+                if (!children_[i]->match(v, captures))
+                {
+                    return false;
+                }
+                ++i;
+            }
+
+            return true;
+        }
+
+        // TODO(QAT-private-issue-33): value may be other type than llvm::User. Check other relevant types
+        // and deal with it.
+
+        return true;
+    }
+
+    void IOperandPrototype::addChild(Child const& child)
+    {
+        children_.push_back(child);
+    }
+
+    void IOperandPrototype::enableCapture(std::string capture_name)
+    {
+        capture_name_ = std::move(capture_name);
+    }
+
+    bool IOperandPrototype::fail(Value* /*value*/, Captures& /*captures*/) const
+    {
+        return false;
+    }
```

Comment:
Why would one need a non-virtual member function that always returns `false`?

Reply:
This is aesthetics and future-proofing the design: Any match function implementation should return either `success` or `fail`, rather than having the options to be `success()` and `false`. This creates symmetry in the return pattern and ensures that if we in the future need to make updates on the basis of a pattern failing, we can easily do that.

Reply:
(I can be wrong but)
As long as function `fail()` is _not virtual_, its behavior and the mechanisms of those calling it, becomes some sort of set in stone (always do nothing, just return `false`. Just a redundant call to `fail()`). But if `fail()` is made _virtual_ then the behavior becomes flexible/extendable.

Reply:
Yes, you are right that it is a redundant call. The point though is: 
1. the ability to add failure code at a later point without having to change every operand defined
2. making the code "symmetric", i.e.
```
    bool PhiPattern::match(Value* instr, Captures& captures) const
    {
        auto* phi_node = llvm::dyn_cast<llvm::PHINode>(instr);
        if (phi_node == nullptr)
        {
            return fail(instr, captures);
        }

        return success(instr, captures);
    }
```
instead of
```
    bool PhiPattern::match(Value* instr, Captures& captures) const
    {
        auto* phi_node = llvm::dyn_cast<llvm::PHINode>(instr);
        if (phi_node == nullptr)
        {
            return false;
        }

        return success(instr, captures);
    }
```